platform/freedesktop/display_wayland.h

```diff
@@ -0,0 +1,225 @@
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+/**
+	@author Juan Linietsky <reduzio@gmail.com>
+*/
+#include "core/os/displaydriver.h"
+#include "core/os/input.h"
+#include "joypad_linux.h"
+#include "main/input_default.h"
+#include "power_x11.h"
+#include "servers/visual/rasterizer.h"
+#include "servers/visual_server.h"
+
+#include <wayland-client-protocol.h>
+#include <wayland-client.h>
+#include <wayland-egl.h> // Wayland EGL MUST be included before EGL headers
+#include <wayland-server.h>
+
+#include "context_egl_wayland.h"
+#include "wayland_protocol/xdg-shell.h"
+#undef CursorShape
+
+class Display_wayland : public DisplayDriver {
+private:
+	MainLoop *main_loop;
+	VisualServer *visual_server;
+	VideoMode current_videomode;
+	List<String> args;
+	ContextGL_EGL *context_gl_egl;
+	struct wl_compositor *compositor = NULL;
+	struct wl_surface *surface;
+	struct wl_egl_window *egl_window;
+	struct wl_region *region;
+	// struct wl_shell *shell;
+	struct zxdg_shell_v6 *xdg_shell;
+	struct zxdg_surface_v6 *xdg_surface;
+	struct zxdg_toplevel_v6 *xdg_toplevel;
+	//wayland listeners
+	const struct wl_registry_listener registry_listener = {
+		&global_registry_handler,
+		&global_registry_remover
+	};
+	static void global_registry_handler(void *data, struct wl_registry *registry, uint32_t id, const char *interface, uint32_t version);
+	static void global_registry_remover(void *data, struct wl_registry *wl_registry, uint32_t name);
+
+	const struct zxdg_toplevel_v6_listener xdg_toplevel_listener = {
+		.configure = &xdg_toplevel_configure_handler,
+		.close = &xdg_toplevel_close_handler
+	};
+	static void xdg_toplevel_configure_handler(void *data, struct zxdg_toplevel_v6 *xdg_toplevel, int32_t width, int32_t height, struct wl_array *states);
+	static void xdg_toplevel_close_handler(void *data, struct zxdg_toplevel_v6 *xdg_toplevel);
+
+	const struct zxdg_surface_v6_listener xdg_surface_listener = {
+		.configure = &xdg_surface_configure_handler
+	};
+	static void xdg_surface_configure_handler(void *data, struct zxdg_surface_v6 *xdg_surface, uint32_t serial);
+
+	const struct zxdg_shell_v6_listener xdg_shell_listener = {
+		.ping = &xdg_shell_ping_handler
+	};
+	static void xdg_shell_ping_handler(void *data, struct zxdg_shell_v6 *xdg_shell, uint32_t serial);
```

Comment:
good idea. keep the header clutter free since there will be A LOT of fucntions...
maybe than i just use a `#define disp_driver Display_wayland::get_singleton()`

Reply:
> `#define disp_driver Display_wayland::get_singleton()`

What, what? No no no no no. Plz don't do that.